File: enjoymadrid-backend/src/main/java/com/example/enjoymadrid/services/TouristicPointService.java
package com.example.enjoymadrid.services;

import java.util.List;

import com.example.enjoymadrid.models.TouristicPoint;

public interface TouristicPointService {

	/**
	 * Get tourist points associated to a category
	 * 
	 * @param category Category of tourist points
	 * @return Tourist points
	 */
	public List<TouristicPoint> getTouristicPointsByCategory(String category);
	
	/**
	 * Get tourist points in order of relevance by a means of a query
	 * 
	 * @param query Query to rank tourist points
	 * @return Tourist points
	 */
	public List<TouristicPoint> getTouristicPointsByQuery(String query);
	
	/**
	 * Get the tourist points associated to a user
	 * 
	 * @param userId ID of a user
	 * @return Tourist points
	 */
	public List<TouristicPoint> getUserTouristicPoints(Long userId);
	
	/**
	 * Add a tourist point to a user
	 * 
	 * @param userId ID of a user
	 * @param touristPointId ID of a tourist point
	 */
	public void addTouristicPointToUser(Long userId, Long touristPointId);
	
	/**
	 * Unbundled a tourist point from a user
	 * 
	 * @param userId ID of a user
	 * @param touristPointId ID of a tourist point
	 */
	public void deleteUserTouristicPoint(Long userId, Long touristPointId);
	
}

File: enjoymadrid-frontend/src/app/services/touristic-point/touristic-point.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { TouristicPointModel } from 'src/app/models/touristic-point.model';
import { SharedService } from '../shared/shared.service';

@Injectable({
  providedIn: 'root'
})
export class TouristicPointService {

  constructor(
    private httpClient: HttpClient, 
    private sharedService: SharedService
  ) { }

  getTouristicPointsByCategory(category: string): Observable<TouristicPointModel[]> {
    return this.httpClient.get<TouristicPointModel[]>(this.sharedService.getApiUrl() + 'tourist-points?category=' + category).pipe(
      catchError(this.sharedService.handleError)
    );
  }

  getTouristicPointsByQuery(query: string): Observable<TouristicPointModel[]> {
    return this.httpClient.get<TouristicPointModel[]>(this.sharedService.getApiUrl() + 'tourist-points?query=' + query).pipe(
      catchError(this.sharedService.handleError)
    );
  }

  getUserTouristicPoints(userId: number): Observable<TouristicPointModel[]> {
    return this.httpClient.get<TouristicPointModel[]>(this.sharedService.getApiUrl() + 'users/' + userId + '/tourist-points' ).pipe(
      catchError(this.sharedService.handleError)
    );
  }

  addTouristicPointToUser(userId: number, touristPointId: number): Observable<any> {
    return this.httpClient.post<any>(this.sharedService.getApiUrl() + 'users/' + userId + '/tourist-points/' + touristPointId, []).pipe(
      catchError(this.sharedService.handleError)
    );
  } 

  deleteUserTouristicPoint(userId: number, touristPointId: number): Observable<any> {
    return this.httpClient.delete<any>(this.sharedService.getApiUrl() + 'users/' + userId + '/tourist-points/' + touristPointId).pipe(
      catchError(this.sharedService.handleError)
    );
  }

}
